### PR TITLE
tracing: add gpio tracing capabilities and sample

### PR DESCRIPTION
**Pull request by alexanderlayTT**
If GPIO tracing is enabled, then the system will track input, output, interrupt attachment/removal and interrupt execution for GPIO pins.

### CODE REVIEW

File: include/zephyr/tracing/tracing.h [hunk 1]
```diff
@@ -2324,6 +2324,65 @@
...
+ * @param port @ref device representing the GPIO port
+ * @param pin @ref gpio_pin_t representing the pin
+ * @param flags @ref gpio_flags_t representing the flags
+ */
+#define sys_port_trace_gpio_pin_configured_input(port, pin, flags)
+
+/**
+ * @brief Trace attaching event handler to GPIO
+ * @param port @ref device representing the GPIO port
+ * @param callback @ref a valid Application's callback structure pointer
```
Comment (cfriedt) [inline include/zephyr/tracing/tracing.h:2366]:
```suggestion
 * @param callback pointer to @ref gpio_callback
```

File: include/zephyr/tracing/tracing.h [hunk 2]
```diff
@@ -2324,6 +2324,65 @@
...
+ */
+#define sys_port_trace_gpio_pin_event_attached(port, callback)
+
+/**
+ * @brief Trace removing event handler from GPIO
+ * @param port @ref device representing the GPIO port
+ * @param callback @ref a valid Application's callback structure pointer
```
Comment (cfriedt) [inline include/zephyr/tracing/tracing.h:2373]:
```suggestion
 * @param callback pointer to @ref gpio_callback
```

File: include/zephyr/tracing/tracing.h [hunk 3]
```diff
@@ -2324,6 +2324,65 @@
...
+ */
+#define sys_port_trace_gpio_pin_event_removed(port, callback)
+
+/**
+ * @brief Trace executing GPIO event handler
+ * @param port @ref device representing the GPIO port
+ * @param callback @ref a valid Application's callback structure pointer
```
Comment (cfriedt) [inline include/zephyr/tracing/tracing.h:2380]:
```suggestion
 * @param callback pointer to @ref gpio_callback
```

**Comment by cfriedt** [comment]
The warnings about braces being on a newline are false positives (it's following the existing convention in the `tracing_user.c`).

File: samples/subsys/tracing/src/gpio_main.c [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+	/* Configure pin 1 as input */
+	gpio_pin_configure(dev, 1, GPIO_INPUT);
+
+	/* Read pin 1 */
+	if (IS_ENABLED(CONFIG_DT_HAS_GPIO_EMUL_ENABLED)) {
+		gpio_emul_input_set(dev, 1, 1);
+	}
+	gpio_pin_get(dev, 1);
+	if (IS_ENABLED(CONFIG_DT_HAS_GPIO_EMUL_ENABLED)) {
```
Comment (cfriedt) [inline samples/subsys/tracing/src/gpio_main.c:44]:
```suggestion
	if (DT_HAS_COMPAT_STATUS_OKAY(zephyr_gpio_emul)) {
```

File: samples/subsys/tracing/src/gpio_main.c [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+	/* Configure pin 0 as output and toggle */
+	gpio_pin_configure(dev, 0, GPIO_OUTPUT);
+	gpio_pin_set(dev, 0, 1);
+	gpio_pin_set(dev, 0, 0);
+
+	/* Configure pin 1 as input */
+	gpio_pin_configure(dev, 1, GPIO_INPUT);
+
+	/* Read pin 1 */
+	if (IS_ENABLED(CONFIG_DT_HAS_GPIO_EMUL_ENABLED)) {
```
Comment (cfriedt) [inline samples/subsys/tracing/src/gpio_main.c:40]:
```suggestion
	if (DT_HAS_COMPAT_STATUS_OKAY(zephyr_gpio_emul)) {
```

File: samples/subsys/tracing/src/gpio_main.c [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+	/* Setup pin 1 for interrupt */
+	gpio_pin_interrupt_configure(dev, 1, GPIO_INT_EDGE_RISING);
+	static struct gpio_callback gpio_cb;
+
+	gpio_init_callback(&gpio_cb, test_handler, BIT(1));
+	gpio_add_callback(dev, &gpio_cb);
+
+	/* Trigger interrupt */
+	if (IS_ENABLED(CONFIG_DT_HAS_GPIO_EMUL_ENABLED)) {
```
Comment (cfriedt) [inline samples/subsys/tracing/src/gpio_main.c:57]:
```suggestion
	if (DT_HAS_COMPAT_STATUS_OKAY(zephyr_gpio_emul)) {
```

**Review by cfriedt — DISMISSED** [review]

**Comment by cfriedt** [comment]
Action shot (heavily customized to use [Chrome Trace JSON](https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/preview#heading=h.yr4qxyxotyw) output) where the trace subsys was used to trace bit-banging JTAG + AXI.

Screenshot 2024-05-31 at 5 20 30 PM

**Comment by cfriedt** [comment]
@alexanderlayTT - can you fix the compliance checks and rebase on `main`?

```
OPEN_BRACE: open brace '{' following function definitions go on the next line
File:subsys/tracing/user/tracing_user.c
Line:30
 OPEN_BRACE: open brace '{' following function definitions go on the next line
File:subsys/tracing/user/tracing_user.c
Line:32
 OPEN_BRACE: open brace '{' following function definitions go on the next line
File:subsys/tracing/user/tracing_user.c
Line:[34](https://github.com/zephyrproject-rtos/zephyr/actions/runs/9649686085/job/26622907424?pr=74199#step:10:35)
 OPEN_BRACE: open brace '{' following function definitions go on the next line
File:subsys/tracing/user/tracing_user.c
Line:[36](https://github.com/zephyrproject-rtos/zephyr/actions/runs/9649686085/job/26622907424?pr=74199#step:10:37)
 OPEN_BRACE: open brace '{' following function definitions go on the next line
File:subsys/tracing/user/tracing_user.c
Line:[38](https://github.com/zephyrproject-rtos/zephyr/actions/runs/9649686085/job/26622907424?pr=74199#step:10:39)
```

File: include/zephyr/drivers/gpio.h [hunk 7]
```diff
@@ -1236,7 +1248,25 @@ static inline int z_impl_gpio_port_get_raw(const struct device *port,
...
+	if (ret != 0) {
+		return ret;
+	}
+
+	int counter = 0;
+	gpio_port_value_t val_cp = *value;
+
+	while (val_cp != 0) {
+		if ((val_cp & 0x1) == 1) {
+			sys_port_trace_gpio_pin_active(port, (gpio_pin_t)counter);
```
Comment (ycsin) [inline include/zephyr/drivers/gpio.h:1262]: I wonder if the compiler optimizes out this loop if tracing is disabled?

File: include/zephyr/drivers/gpio.h [hunk 8]
```diff
@@ -1236,7 +1248,25 @@ static inline int z_impl_gpio_port_get_raw(const struct device *port,
...
 		(const struct gpio_driver_api *)port->api;
 
-	return api->port_get_raw(port, value);
+	int ret = api->port_get_raw(port, value);
+
+	if (ret != 0) {
+		return ret;
+	}
+
+	int counter = 0;
```
Comment (ycsin) [inline include/zephyr/drivers/gpio.h:1257]: probably could have declare it as `gpio_pin_t` to prevent the casting later

File: include/zephyr/drivers/gpio.h [hunk 9]
```diff
@@ -1300,6 +1330,16 @@ static inline int z_impl_gpio_port_set_masked_raw(const struct device *port,
 	const struct gpio_driver_api *api =
 		(const struct gpio_driver_api *)port->api;
 
+	for (size_t i = 0; i < sizeof(mask) * 8; ++i) {
```
Comment (ycsin) [inline include/zephyr/drivers/gpio.h:1333]: probably could have declare `i` as `gpio_pin_t` to prevent the casting later

File: samples/subsys/tracing/src/tracing_user.c [hunk 10]
```diff
@@ -56,3 +58,43 @@ void sys_trace_idle_user(void)
...
+	printk("port: %s, pin: %d status: active\n", port->name, pin);
+}
+
+void sys_trace_gpio_pin_inactive_user(const struct device *port, gpio_pin_t pin)
+{
+	printk("port: %s, pin: %d status: inactive\n", port->name, pin);
+}
+
+void sys_trace_gpio_pin_configured_output_user(const struct device *port, gpio_pin_t pin,
+						gpio_flags_t flags)
```
Comment (ycsin) [inline samples/subsys/tracing/src/tracing_user.c:73]: this and the following are not aligned

File: samples/subsys/tracing/src/tracing_user.c [hunk 11]
```diff
@@ -4,7 +4,9 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-#include <tracing_user.h>
+#include "tracing_user.h"
```
Comment (ycsin) [inline samples/subsys/tracing/src/tracing_user.c:7]: this line probably can be left unchanged
```suggestion
#include <tracing_user.h>
```

File: samples/subsys/tracing/src/gpio_main.c [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+
+
+/* size of stack area used by each thread */
+#define STACKSIZE (2048)
+
+/* scheduling priority used by each thread */
+#define PRIORITY 7
+
+
+void test_handler(const struct device *port, struct gpio_callback *cb, uint32_t pins)
```
Comment (ycsin) [inline samples/subsys/tracing/src/gpio_main.c:21]:
```suggestion
void test_handler(const struct device *port, struct gpio_callback *cb, gpio_port_pins_t pins)
```

File: samples/subsys/tracing/sample.yaml [hunk 13]
```diff
@@ -2,16 +2,19 @@ common:
   tags:
     - tracing
     - debug
-  harness: console
+    - gpio
+  depends_on: gpio
+  harness: ztest
```
Comment (ycsin) [inline samples/subsys/tracing/sample.yaml:7]: The original testcase tests the tracing in a generic way, this changed that behavior and probably isn't ideal. You probably want to do this (gpio-tracing related) in the `sample.tracing.gpio`, or create another testcase if that isn't possible, and leave the other testcases alone

File: samples/subsys/tracing/src/gpio_main.c [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+	if (DT_HAS_COMPAT_STATUS_OKAY(zephyr_gpio_emul)) {
+		gpio_emul_input_set(dev, 1, 1);
+	}
+
+	/* Remove interrupt */
+	gpio_remove_callback(dev, &gpio_cb);
+}
+
+
+void threadGPIO(void *dummy1, void *dummy2, void *dummy3)
```
Comment (ycsin) [inline samples/subsys/tracing/src/gpio_main.c:66]:
```suggestion
static void gpio_thread(void *dummy1, void *dummy2, void *dummy3)
```

File: samples/subsys/tracing/src/gpio_main.c [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+{
+	ARG_UNUSED(dummy1);
+	ARG_UNUSED(dummy2);
+	ARG_UNUSED(dummy3);
+
+	gpio_sample();
+}
+
+
+K_THREAD_DEFINE(thread_gpio, STACKSIZE, threadGPIO, NULL, NULL, NULL,
```
Comment (ycsin) [inline samples/subsys/tracing/src/gpio_main.c:76]:
```suggestion
K_THREAD_DEFINE(thread_gpio, STACKSIZE, gpio_thread, NULL, NULL, NULL,
```

File: samples/subsys/tracing/src/gpio_main.c [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+
+void test_handler(const struct device *port, struct gpio_callback *cb, uint32_t pins)
+{
+	printk("Interrupt detected!\n");
+}
+
+
+void gpio_sample(void)
+{
+	const struct device *dev = DEVICE_DT_GET(DT_NODELABEL(gpio0));
```
Comment (ycsin) [inline samples/subsys/tracing/src/gpio_main.c:29]: You can create a unique gpio-emul node in `app.overlay` so that this can work in every boards and don't have to rely on native-sim

File: samples/subsys/tracing/src/gpio_main.c [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+	/* Configure pin 0 as output and toggle */
+	gpio_pin_configure(dev, 0, GPIO_OUTPUT);
+	gpio_pin_set(dev, 0, 1);
+	gpio_pin_set(dev, 0, 0);
+
+	/* Configure pin 1 as input */
+	gpio_pin_configure(dev, 1, GPIO_INPUT);
+
+	/* Read pin 1 */
+	if (DT_HAS_COMPAT_STATUS_OKAY(zephyr_gpio_emul)) {
```
Comment (ycsin) [inline samples/subsys/tracing/src/gpio_main.c:40]: `DT_HAS_COMPAT_STATUS_OKAY()` checks if there's any `zephyr_gpio_emul` that has status "okay" in the devicetree, but not necessarily the one that you are using (`gpio0`).

`gpio0` must have status "okay", otherwise `DEVICE_DT_GET(DT_NODELABEL(gpio0))` fails to compile.

To check if `gpio0` has been initialized properly, you can do `device_is_ready(dev)` check after L29.

The repeated `if (DT_HAS_COMPAT_STATUS_OKAY(zephyr_gpio_emul))` checks are redundant after that.

**Review by ycsin — CHANGES_REQUESTED** [review]

File: include/zephyr/drivers/gpio.h (hunk 7 above)
Comment (cfriedt) [reply include/zephyr/drivers/gpio.h:1262]: Probably could surround it with `if (IS_ENABLED(CONFIG_TRACING_GPIO)) { ... }`

File: include/zephyr/drivers/gpio.h (hunk 9 above)
Comment (cfriedt) [reply include/zephyr/drivers/gpio.h:1333]: Also maybe a good idea to wrap with `if (IS_ENABLED(CONFIG_TRACING_GPIO)) { ... }`

File: include/zephyr/drivers/gpio.h [hunk 18]
```diff
@@ -1236,7 +1248,25 @@ static inline int z_impl_gpio_port_get_raw(const struct device *port,
...
+	int ret = api->port_get_raw(port, value);
+
+	if (ret != 0) {
+		return ret;
+	}
+
+	int counter = 0;
+	gpio_port_value_t val_cp = *value;
+
+	while (val_cp != 0) {
```
Comment (cfriedt) [inline include/zephyr/drivers/gpio.h:1260]:
```suggestion
	for (gpio_port_value_t val_cp = *value; IS_ENABLED(CONFIG_TRACING_GPIO) && (val_cp != 0); val_cp >>= 1, counter++) {
```

File: include/zephyr/drivers/gpio.h [hunk 19]
```diff
@@ -1236,7 +1248,25 @@ static inline int z_impl_gpio_port_get_raw(const struct device *port,
...
+		if ((val_cp & 0x1) == 1) {
+			sys_port_trace_gpio_pin_active(port, (gpio_pin_t)counter);
+		}
+
+		val_cp >>= 1;
+		counter++;
```
Comment (cfriedt) [inline include/zephyr/drivers/gpio.h:1266]: these lines can be removed if the for loop is used.

File: include/zephyr/drivers/gpio.h (hunk 9 above)
Comment (cfriedt) [inline include/zephyr/drivers/gpio.h:1333]:
```suggestion
	for (size_t i = 0; IS_ENABLED(CONFIG_TRACING_GPIO) && (i < sizeof(mask) * 8); ++i) {
```

File: include/zephyr/drivers/gpio.h [hunk 20]
```diff
@@ -1230,13 +1242,27 @@ static inline int gpio_pin_get_config_dt(const struct gpio_dt_spec *spec,
...
+
+	int ret = api->port_get_raw(port, value);
+
+	if (ret != 0) {
+		return ret;
+	}
 
-	return api->port_get_raw(port, value);
+	gpio_pin_t counter = 0;
+	gpio_port_value_t val_cp = *value;
```
Comment (ycsin) [inline include/zephyr/drivers/gpio.h:1256]: This line is now redundant

```suggestion
```

File: subsys/tracing/user/tracing_user.c [hunk 21]
```diff
@@ -105,3 +117,40 @@ void sys_trace_sys_init_exit(const struct init_entry *entry, int level, int resu
...
+							gpio_flags_t flags)
+{
+	sys_trace_gpio_pin_configured_output_user(port, pin, flags);
+}
+
+void sys_trace_gpio_pin_configured_input(const struct device *port, gpio_pin_t pin,
+							gpio_flags_t flags)
+{
+	sys_trace_gpio_pin_configured_input_user(port, pin, flags);
+}
```
Comment (ycsin) [inline subsys/tracing/user/tracing_user.c:141]: align

**Review by ycsin — CHANGES_REQUESTED** [review]

File: samples/subsys/tracing/sample.yaml [hunk 22]
```diff
@@ -8,10 +8,11 @@ common:
     regex:
       - "thread_a: Hello World from (.*)!"
       - "thread_b: Hello World from (.*)!"
+    fixture: gpio_loopback
 tests:
   sample.tracing.user:
     extra_args: CONF_FILE="prj_user.conf"
-    integration_platforms:
+    platform_allow:
```
Comment (ycsin) [inline samples/subsys/tracing/sample.yaml:15]: why do you need to change these testcases?

**Review by cfriedt — DISMISSED** [review]
LGTM

File: subsys/tracing/ctf/tracing_ctf.h
Comment (kartben) [inline subsys/tracing/ctf/tracing_ctf.h:1]: I might be wrong but I believe `subsys/tracing/ctf/tsdl/metadata` also requires updating?

File: samples/subsys/tracing/prj_percepio.conf [hunk 23]
```diff
@@ -4,3 +4,4 @@ CONFIG_MP_MAX_NUM_CPUS=1
 CONFIG_DEBUG_THREAD_INFO=y
 CONFIG_PERCEPIO_TRACERECORDER=y
 CONFIG_TRACING=y
+CONFIG_TRACING_GPIO=n
```
Comment (ycsin) [inline samples/subsys/tracing/prj_percepio.conf:7]: why are these required? wouldn't they just compile to nothing if they are not implemented by this backends?

Comment (alexanderlayTT) [reply samples/subsys/tracing/prj_percepio.conf:7]: Yeah that makes sense - we will fix that

File: samples/subsys/tracing/app.overlay [hunk 24]
```diff
@@ -0,0 +1,19 @@
+/*
+ * Copyright (c) 2024 Tenstorrent AI ULC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/ {
+    gpio_emul: gpio_emul {
```
Comment (cfriedt) [inline samples/subsys/tracing/app.overlay:8]: Indentation looks off a bit here. Probably 4 spaces that need to be replaced by a tab.

File: include/zephyr/drivers/gpio.h [hunk 25]
```diff
@@ -550,7 +551,7 @@ struct gpio_dt_spec {
  *		.ngpios = 32,
  *		.gpios_reserved = 0xdeadbeef,
  *		               // 0b1101 1110 1010 1101 1011 1110 1110 1111
- *
+ *	};
```
Comment (cfriedt) [inline include/zephyr/drivers/gpio.h:554]: Since this 1-line update to a comment isn't specifically related to the tracing change, it might be better as a separate commit (i realize it's a lot of overhead for a 1-line change, but helps with change management in the long run).

**Review by cfriedt — DISMISSED** [review]

File: samples/subsys/tracing/prj_percepio.conf (hunk 23 above)
Comment (cfriedt) [reply samples/subsys/tracing/prj_percepio.conf]: @ycsin - does this look OK now?

File: samples/subsys/tracing/sample.yaml (hunk 22 above)
Comment (cfriedt) [reply samples/subsys/tracing/sample.yaml]: @ycsin - does this look ok now?

File: samples/subsys/tracing/sample.yaml [hunk 26]
```diff
@@ -8,6 +8,7 @@ common:
     regex:
       - "thread_a: Hello World from (.*)!"
       - "thread_b: Hello World from (.*)!"
+  extra_args: EXTRA_DTC_OVERLAY_FILE="app.overlay"
```
Comment (ycsin) [inline samples/subsys/tracing/sample.yaml:11]: could you please check if this can be moved to `sample.tracing.gpio` so it affects that test case only? It should look something like:

```yaml
    extra_args:
      - CONF_FILE=prj_gpio.conf
      - DTC_OVERLAY_FILE=gpio.overlay
```

File: samples/subsys/tracing/app.overlay [hunk 27]
```diff
@@ -0,0 +1,19 @@
+/*
+ * Copyright (c) 2024 Tenstorrent AI ULC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/	{
+    gpio_emul: gpio_emul {
```
Comment (ycsin) [inline samples/subsys/tracing/app.overlay:8]:
```suggestion
/ {
	gpio_emul: gpio_emul {
```

File: samples/subsys/tracing/src/gpio_main.c [hunk 28]
```diff
@@ -0,0 +1,76 @@
...
+static void gpio_thread(void *dummy1, void *dummy2, void *dummy3)
+{
+	ARG_UNUSED(dummy1);
+	ARG_UNUSED(dummy2);
+	ARG_UNUSED(dummy3);
+
+	gpio_sample();
+}
+
+
```
Comment (ycsin) [inline samples/subsys/tracing/src/gpio_main.c:74]: nit: looks somewhat from pep8

**Review by ycsin — CHANGES_REQUESTED** [review]
I tried to rerun the compliance check but it still fails, I guess this probably needs a rebase, in that case I think we can grab the opportunity to fix some nits, LGTM otherwise

**Comment by alexanderlayTT** [comment]
@cfriedt @ycsin I believe we have addressed all the changes, PTAL

**Review by cfriedt — DISMISSED** [review]

File: subsys/tracing/user/tracing_user.c [hunk 29]
```diff
@@ -25,6 +25,18 @@ void __weak sys_trace_isr_exit_user(void) {}
 void __weak sys_trace_idle_user(void) {}
 void __weak sys_trace_sys_init_enter_user(const struct init_entry *entry, int level) {}
 void __weak sys_trace_sys_init_exit_user(const struct init_entry *entry, int level, int result) {}
+void __weak sys_trace_gpio_pin_active_user(const struct device *port, gpio_pin_t pin) {}
+void __weak sys_trace_gpio_pin_inactive_user(const struct device *port, gpio_pin_t pin) {}
+void __weak sys_trace_gpio_pin_configured_output_user(const struct device *port, gpio_pin_t pin,
```
Comment (ycsin) [inline subsys/tracing/user/tracing_user.c:30]: @nashif any idea about these compliance errors? do we have to follow the recommendation but break consistency with the existing declarations?

File: samples/subsys/tracing/app.overlay [hunk 30]
```diff
@@ -0,0 +1,19 @@
+/*
```
Comment (ycsin) [inline samples/subsys/tracing/app.overlay:1]: I guess this should be renamed to something else, i.e. `gpio.overlay`, as `app.overlay` has special meaning and will be added by the build system [by default](https://docs.zephyrproject.org/latest/develop/application/index.html#overview)

Comment (cfriedt) [reply samples/subsys/tracing/app.overlay:1]: That shouldn't really be a huge concern though, for a sample, should it? 

Alexander did add app.overlay earlier due to this comment
https://github.com/zephyrproject-rtos/zephyr/pull/74199#discussion_r1697269111

Comment (ycsin) [reply samples/subsys/tracing/app.overlay:1]: > Alexander did add app.overlay earlier due to this comment
> [#74199 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/74199#discussion_r1697269111)

The gpio test is now completely contained in a testcase of its own, so this looks like the final piece of a 1k pieces puzzle to me

> That shouldn't really be a huge concern though, for a sample, should it?

This sample doubles as a test, but yeah, it is not going to break anything so I'm going to stamp it

**Review by ycsin — DISMISSED** [review]
Very good work, but the compliance error has to be fixed or merged by superpower

File: include/zephyr/drivers/gpio.h [hunk 31]
```diff
@@ -1230,13 +1242,26 @@ static inline int gpio_pin_get_config_dt(const struct gpio_dt_spec *spec,
...
+
+	gpio_pin_t counter = 0;
 
-	return api->port_get_raw(port, value);
+	for (gpio_port_value_t val_cp = *value; IS_ENABLED(CONFIG_TRACING_GPIO) && (val_cp != 0);
+	     val_cp >>= 1, counter++) {
+		if ((val_cp & 0x1) == 1) {
+			sys_port_trace_gpio_pin_active(port, counter);
+		}
+	}
```
Comment (bjarki-andreasen) [inline include/zephyr/drivers/gpio.h:1262]: Why not add tracing for `port_get_raw()`? this conversion from mask to individual pin events will likely take even longer than the call to the API itself

Comment (bjarki-andreasen) [reply include/zephyr/drivers/gpio.h:1262]: also, it does not seem to respect the ACTIVE_HIGH/ACTIVE_LOW flags, which determine if a 1 or 0 means `active` or not, which it seems is what is being traced by `sys_port_trace_gpio_pin_active()` rather than the raw value.

File: include/zephyr/tracing/tracing.h [hunk 32]
```diff
@@ -2376,6 +2376,65 @@
...
+ * @param callback @ref gpio_callback a valid Application's callback structure pointer
+ */
+#define sys_port_trace_gpio_pin_event_removed(port, callback)
+
+/**
+ * @brief Trace executing GPIO event handler
+ * @param port @ref device representing the GPIO port
+ * @param callback @ref gpio_callback a valid Application's callback structure pointer
+ */
+#define sys_port_trace_gpio_pin_event_executed(port, callback)
```
Comment (bjarki-andreasen) [inline include/zephyr/tracing/tracing.h:2434]: this event is not a `pin` event, there is not even a mask of pins, maybe just name it `sys_port_trace_gpio_callback_called()`? The same comment applies to `sys_port_trace_gpio_pin_event_attached()` and `sys_port_trace_gpio_pin_event_removed()`, where you may want to use `added` and `removed` respectively to match the wording of the GPIO API :)

File: include/zephyr/tracing/tracing.h [hunk 33]
```diff
@@ -2376,6 +2376,65 @@
...
+ * @defgroup subsys_tracing_apis_gpio GPIO Tracing APIs
+ * @{
+ */
+
+/**
+ * @brief Trace a GPIO pin becoming active
+ * @param port @ref device representing the GPIO port
+ * @param pin @ref gpio_pin_t representing the pin
+ */
+#define sys_port_trace_gpio_pin_active(port, pin)
```
Comment (bjarki-andreasen) [inline include/zephyr/tracing/tracing.h:2390]: As the tracing is implemented now in this PR, its not reflective of the GPIO API where `active` is not equal to `high`. The value read using `_raw()` variants of the API return `high` and `low`, not `active` and `inactive` which depend on the configuration of the pins

File: include/zephyr/tracing/tracing.h [hunk 34]
```diff
@@ -2376,6 +2376,65 @@
...
+ */
+#define sys_port_trace_gpio_pin_inactive(port, pin)
+
+/**
+ * @brief Trace GPIO being configured as output
+ * @param port @ref device representing the GPIO port
+ * @param pin @ref gpio_pin_t representing the pin
+ * @param flags @ref gpio_flags_t representing the flags
+ */
+#define sys_port_trace_gpio_pin_configured_output(port, pin, flags)
```
Comment (bjarki-andreasen) [inline include/zephyr/tracing/tracing.h:2405]: Since you are passing the flags here, I don't see a reason to split this trace into two traces (`sys_port_trace_gpio_pin_configured_output` and `sys_port_trace_gpio_pin_configured_input`) since the output/input config is already encoded in the flags.

**Review by bjarki-andreasen — DISMISSED** [review]
I believe the tracing in this PR can be tuned to match the GPIO API better. GPIO controllers work on entire ports at a time, converting this to multiple individual calls to specific pins is quite a bit of overhead and complexity added rather than just tracing the actual API calls like `gpio_port_set_masked_raw()`

That is what we are tracing, right? the number of calls with which parameters to the GPIO controllers? It seems like from this PR we are using tracing to monitor the states of individual GPIO pins, which is not exactly what tracing is for right?

We are not monitoring state changes on GPIO inputs, but, we when the GPIO get API is called, tracing is used to print the state of each individual PIN, seems incorrect IMO.

**Review by nashif — DISMISSED** [review]
> I believe the tracing in this PR can be tuned to match the GPIO API better. GPIO controllers work on entire ports at a time, converting this to multiple individual calls to specific pins is quite a bit of overhead and complexity added rather than just tracing the actual API calls like gpio_port_set_masked_raw()


completely agree with @bjarki-andreasen.

Tracing needs to be close to zero overhead and needs as much as possible to match the API. You are adding logic in many places making things more complex than they should be. look at how things were implemented in other places, especially the kernel.

**Comment by alexanderlayTT** [comment]
> @alexanderlayTT - just a ping to see if you are making progress.

Yes! We had some other work come up, but we are making progress on this too.

**Comment by cfriedt** [comment]
@alexanderlayTT - looks like there is at least one unique build failure.

https://github.com/zephyrproject-rtos/zephyr/actions/runs/11803255472/job/32880806857#step:11:1

What I usually do is copy paste the first build / runtime failure in the logs



and then edit it to run with `twister` to repro.

```shell
twister -i -p native_sim/native -s samples/basic/blinky/sample.basic.blinky
```

Repeat as necessary 😄

Of course, if you scroll to the end of the buffer, there is also a repro command that you can copy & paste.

**Comment by cfriedt** [comment]
@alexanderlayTT - there is still a doc build issue

```
WARNING: /home/runner/work/zephyr/zephyr/include/zephyr/tracing/tracing.h:2595: unable to resolve reference to 'list' for \ref command
WARNING: /home/runner/work/zephyr/zephyr/include/zephyr/tracing/tracing.h:2595: unable to resolve reference to 'list' for \ref command
```

You should be able to reproduce with this
https://docs.zephyrproject.org/latest/contribute/documentation/generation.html

**Review by cfriedt — DISMISSED** [review]

**Comment by cfriedt** [comment]
Looks good 👍 Thanks for the attention to detail @alexanderlayTT

File: subsys/tracing/user/tracing_user.c [hunk 35]
```diff
@@ -25,6 +25,55 @@ void __weak sys_trace_isr_exit_user(void) {}
 void __weak sys_trace_idle_user(void) {}
 void __weak sys_trace_sys_init_enter_user(const struct init_entry *entry, int level) {}
 void __weak sys_trace_sys_init_exit_user(const struct init_entry *entry, int level, int result) {}
+void __weak sys_trace_gpio_pin_interrupt_configure_enter_user(const struct device *port,
```
Comment (ycsin) [inline subsys/tracing/user/tracing_user.c:28]: Is there a way to make the CI not complain about these?

Comment (cfriedt) [reply subsys/tracing/user/tracing_user.c:28]: Maybe 
```
/* clang-format off */
...
/* clang-format on */
```

File: include/zephyr/drivers/gpio.h (hunk 31 above)
Comment (cfriedt) [reply include/zephyr/drivers/gpio.h]: This should be addressed. Please unresolve if you think it is not.

File: include/zephyr/tracing/tracing.h (hunk 33 above)
Comment (cfriedt) [reply include/zephyr/tracing/tracing.h]: This should be addressed. Please unresolve if you think it is not.

File: include/zephyr/tracing/tracing.h (hunk 34 above)
Comment (cfriedt) [reply include/zephyr/tracing/tracing.h]: This should be addressed. Please unresolve if you think it is not.

**Comment by cfriedt** [comment]
ping @bjarki-andreasen for a re-review

**Review by bjarki-andreasen — DISMISSED** [review]
Looks good, nice work!

File: include/zephyr/drivers/gpio.h [hunk 36]
```diff
@@ -875,8 +876,12 @@ static inline int z_impl_gpio_pin_interrupt_configure(const struct device *port,
 		(const struct gpio_driver_data *)port->data;
 	enum gpio_int_trig trig;
 	enum gpio_int_mode mode;
+	int ret;
+
+	sys_port_trace_gpio_pin_interrupt_configure_enter(port, pin, flags);
```
Comment (nashif) [inline include/zephyr/drivers/gpio.h:881]: why not use the macros? for example:

`SYS_PORT_TRACING_OBJ_FUNC_ENTER(gpio_pin_interrupt_configure, pin, flags)`
?

Comment (alexanderlayTT) [reply include/zephyr/drivers/gpio.h]: Pushed a change to address this - PTAL

**Review by bjarki-andreasen — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

File: include/zephyr/drivers/gpio.h [hunk 37]
```diff
@@ -1230,13 +1255,16 @@ static inline int gpio_pin_get_config_dt(const struct gpio_dt_spec *spec,
 __syscall int gpio_port_get_raw(const struct device *port,
 				gpio_port_value_t *value);
 
-static inline int z_impl_gpio_port_get_raw(const struct device *port,
-					   gpio_port_value_t *value)
+static inline int z_impl_gpio_port_get_raw(const struct device *port, gpio_port_value_t *value)
```
Comment (ycsin) [inline include/zephyr/drivers/gpio.h:1258]: such formatting changes should be in a separate commit

**Review by ycsin — APPROVED** [review]

**Review by nashif — APPROVED** [review]
